modules/frontend/querysharding.go

```diff
@@ -38,36 +41,156 @@ func ShardingWare(queryShards int, logger log.Logger) Middleware {
 }
 
 type shardQuery struct {
-	next            Handler
+	next            http.RoundTripper
 	queryShards     int
 	logger          log.Logger
 	blockBoundaries [][]byte
 }
 
-// Do implements Handler
-func (s shardQuery) Do(r *http.Request) (*http.Response, error) {
+// RoundTrip implements http.RoundTripper
+func (s shardQuery) RoundTrip(r *http.Request) (*http.Response, error) {
 	ctx := r.Context()
 	span, ctx := opentracing.StartSpanFromContext(ctx, "frontend.ShardQuery")
 	defer span.Finish()
 
 	// context propagation
 	r = r.WithContext(ctx)
+	reqs, err := s.buildShardedRequests(r)
+	if err != nil {
+		return nil, err
+	}
+
+	// execute requests
+	wg := sync.WaitGroup{}
+	mtx := sync.Mutex{}
+
+	var overallTrace *tempopb.Trace
+	var overallError error
+	statusCode := http.StatusNotFound
+	statusMsg := "trace not found"
+
+	for _, req := range reqs {
+		wg.Add(1)
+		go func(innerR *http.Request) {
+			defer wg.Done()
+
+			resp, err := s.next.RoundTrip(innerR)
+
+			mtx.Lock()
+			defer mtx.Unlock()
+			if err != nil {
+				overallError = err
+			}
+
+			if shouldQuit(r.Context(), statusCode, overallError) {
+				return
+			}
+
+			// check http error
+			if err != nil {
+				_ = level.Error(s.logger).Log("msg", "error querying proxy target", "url", innerR.RequestURI, "err", err)
+				overallError = err
+				return
+			}
```

Comment:
This if-case is unreachable:
- if `err != nil`, we set `overallError` (line 81)
- if `overallError` is not nil, `shouldQuit` returns true

Reply:
good catch. i'm going to move the log line up to `if err != nil` check before `shouldQuit`